src/jaxsim/api/model.py

```diff
@@ -1927,11 +1929,43 @@ def step(
         ),
     )
 
-    return (
+    data = (
         # Store the new state of the model and the new time.
         data.replace(
             state=state_xf,
             time_ns=t0_ns + jnp.array(dt * 1e9).astype(jnp.uint64),
-        ),
+        )
+    )
+
+    # In case of rigid contact model, update data
+    match model.contact_model:
+        case RigidContacts():
+            with data.switch_velocity_representation(VelRepr.Mixed):
+                W_o_C, W_ȯ_C = js.contact.collidable_point_kinematics(model, data)
+                M = js.model.free_floating_mass_matrix(model, data)
+                J_WC = js.contact.jacobian(model, data)
+                terrain_height = jax.vmap(model.contact_model.terrain.height)(
+                    W_o_C[:, 0], W_o_C[:, 1]
+                )
+                terrain_normal = jax.vmap(model.contact_model.terrain.normal)(
+                    W_o_C[:, 0], W_o_C[:, 1]
+                )
+                inactive_collidable_points, _ = RigidContacts.detect_contacts(
+                    W_o_C=W_o_C,
+                    W_o_dot_C=W_ȯ_C,
+                    terrain_height=terrain_height,
+                    terrain_normal=terrain_normal,
+                )
+                BW_nu_post_impact = RigidContacts.compute_impact_velocity(
+                    data=data,
+                    inactive_collidable_points=inactive_collidable_points,
+                    M=M,
+                    J_WC=J_WC,
+                )
+                data = data.reset_base_velocity(BW_nu_post_impact[0:6])
+                data = data.reset_joint_velocities(BW_nu_post_impact[6:])
+
+    return (
+        data,
```

Comment:
Thanks a lot, this was actually a bug! since a new data is returned inside the context manager and it overwrites the initial data instance, the returned data is always in mixed representation! I need to convert it back to the original representation.